codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
     interval: Duration
+    defaultExpiry: Duration
+    batchSize: int
+    retryDelay: Duration
     timer: Timer
     clock: Clock
-    numberOfBlocksPerInterval: int
-    offset: int
+    trackedFutures: TrackedFutures
+
+  Checkpoint* {.serialize.} = object
+    timestamp*: SecondsSince1970
+    progress*: Natural
+
+  DatasetMetadata* {.serialize.} = object
+    expiry*: SecondsSince1970
+    leavesCount*: Natural
+    manifestsCids*: seq[Cid]
+    checkpoint*: Checkpoint
+
+  MissingKey* = object of CodexError
 
 proc new*(
-    T: type BlockMaintainer,
-    repoStore: RepoStore,
-    interval: Duration,
-    numberOfBlocksPerInterval = 100,
+    T: type DatasetMaintainer,
+    blockStore: BlockStore,
+    metaDs: Datastore,
+    defaultExpiry = DefaultDefaultExpiry,
+    interval = DefaultMaintenanceInterval,
+    batchSize = DefaultBatchSize,
+    retryDelay = DefaultRetryDelay,
     timer = Timer.new(),
-    clock: Clock = SystemClock.new()
-): BlockMaintainer =
-  ## Create new BlockMaintainer instance
+    clock: Clock = SystemClock.new(),
+    trackedFutures = TrackedFutures.new()
+): DatasetMaintainer =
+  ## Create new DatasetMaintainer instance
   ##
   ## Call `start` to begin looking for for expired blocks
   ##
-  BlockMaintainer(
-    repoStore: repoStore,
+  DatasetMaintainer(
+    blockStore: blockStore,
+    metaDs: TypedDatastore.init(metaDs),
+    defaultExpiry: defaultExpiry,
     interval: interval,
-    numberOfBlocksPerInterval: numberOfBlocksPerInterval,
+    batchSize: batchSize,
+    retryDelay: retryDelay,
     timer: timer,
     clock: clock,
-    offset: 0)
+    trackedFutures: trackedFutures)
 
-proc deleteExpiredBlock(self: BlockMaintainer, cid: Cid): Future[void] {.async.} =
-  if isErr (await self.repoStore.delBlock(cid)):
-    trace "Unable to delete block from repoStore"
+proc encode(t: Checkpoint): seq[byte] = serializer.toJson(t).toBytes()
+proc decode(T: type Checkpoint, bytes: seq[byte]): ?!T = T.fromJson(bytes)
 
-proc processBlockExpiration(self: BlockMaintainer, be: BlockExpiration): Future[void] {.async} =
-  if be.expiry < self.clock.now:
-    await self.deleteExpiredBlock(be.cid)
-  else:
-    inc self.offset
+proc encode(t: DatasetMetadata): seq[byte] = serializer.toJson(t).toBytes()
+proc decode(T: type DatasetMetadata, bytes: seq[byte]): ?!T = T.fromJson(bytes)
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  leavesCount: Natural,
+  expiry: SecondsSince1970,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  # Starts tracking expiry of a given dataset
+  #
+
+  trace "Tracking an expiry of a dataset", treeCid, leavesCount, expiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await modify[DatasetMetadata](self.metaDs, key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      var md: DatasetMetadata
+
+      if currDatasetMd =? maybeCurrDatasetMd:
```

Comment:
How do you think a client can be tricked into downloading a "poisoned" dataset?

Reply:
You'd need to trick the client into downloading a rigged CID containing the correct treeCid and the wrong number of leaves. Since we register the dataset metadata before download is complete, there's no way for the client to check that beforehand (AFAIU), and there you have it.

Reply:
Maybe then a solution would be a bisect? So always take the max number of leafs reported (even if it's `int.high`) and then bisect down to where it actually starts.